lib/services/extensions-service-base.ts

```diff
@@ -0,0 +1,97 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+import path = require("path");
+import options = require("../common/options");
+import helpers = require("../helpers");
+
+export class ExtensionsServiceBase {
+	private extensionVersions: IStringDictionary = {};
+
+	constructor(public $fs: IFileSystem,
+				public $httpClient: Server.IHttpClient,
+				public $logger: ILogger) {
+		if (this.$fs.exists(this.versionsFile).wait()) {
+			this.extensionVersions = this.$fs.readJson(this.versionsFile).wait() || {};
+		}
+	}
+
+	public get cacheDir(): string {
+		return path.join(options.profileDir, "Cache");
+	}
+
+	public getExtensionVersion(packageName: string): string {
+		return this.extensionVersions[packageName];
+	}
+
+	public getExtensionPath(packageName: string): string {
+		return path.join(this.cacheDir, packageName);
+	}
+
+	public prepareExtensionBase(extensionData: IExtensionData, initialCachedVersion: string, beforeDownloadAction?: () => IFuture<void>): IFuture<void> {
+		return ((): void => {
+			var packageName = extensionData.packageName;
+			var extensionVersion = extensionData.version;
+			var extensionPath = extensionData.pathToSave || this.getExtensionPath(extensionData.packageName);
+
+			var cachedVersion = initialCachedVersion;
+			this.$logger.debug("Server version: %s", extensionVersion);
+
+			if (this.extensionVersions[packageName]) {
+				cachedVersion = this.extensionVersions[packageName];
+				this.$logger.debug("Cached version is: %s", cachedVersion);
+			}
+
+			if( helpers.versionCompare(cachedVersion, extensionVersion) < 0) {
+				this.$logger.info("Updating %s package...", packageName);
+				var zipFileName = path.join(this.cacheDir, packageName + ".zip");
+
+				if(beforeDownloadAction) {
+					beforeDownloadAction().wait();
+				}
+
+				if(this.$fs.exists(extensionPath).wait()) {
```

Comment:
The if is superfluous and can be removed. See [rimraf](https://www.npmjs.com/package/rimraf)